src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -1109,14 +1109,26 @@ trait Namers extends MethodSynthesis {
           case _ => sym.isStable && !refersToSymbolLessAccessibleThan(tpe, sym)
         }
 
+      def keepWhenMissResultTypeForSingleton: Option[Type] = {
+        sym.overrideChain.drop(1).map(_.tpe).headOption.flatMap {
+          case MethodType(_, rt) if !rt.typeSymbol.isAliasType && rt.isInstanceOf[SingletonType] && tpe <:< rt.dealiasWiden => Some(rt)
```

Comment:
`overrideChain` may be too much; compare `computeOverridden`. What is `isAliasType` guarding? A test would show it.